File: Android-My-Nanny-old-version/app/src/main/java/com/vald3nir/mynanny/views/login/LoginActivity.java
package com.vald3nir.mynanny.views.login;

import android.content.Intent;
import android.os.Bundle;
import android.widget.EditText;

import com.vald3nir.mynanny.R;
import com.vald3nir.mynanny.views.MainActivity;
import com.vald3nir.mynanny.views.custom.CustomActivity;
import com.vald3nir.mynanny.views.register.RegisterActivity;

import static com.vald3nir.mynanny.utils.AppUtils.getValue;


public class LoginActivity extends CustomActivity {

    private LoginDelegate delegate;
    private EditText edtEmail, edtPassword;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);
        delegate = new LoginDelegate(this);
        initViews();
    }

    @Override
    public void onBackPressed() {
    }

    private void initViews() {
        this.edtEmail = findViewById(R.id.edt_email);
        this.edtPassword = findViewById(R.id.edt_password);

        findViewById(R.id.btn_login).setOnClickListener(v -> login());
        findViewById(R.id.btn_register).setOnClickListener(v -> callRegister());
    }

    private void login() {
        delegate.login(this, getValue(edtEmail), getValue(edtPassword));
    }

    public void callMainActivity() {
        startActivity(new Intent(this, MainActivity.class));
    }

    private void callRegister() {
        startActivity(new Intent(this, RegisterActivity.class));
    }
}

File: app/src/main/java/com/vald3nir/my_nanny/presentation/login/LoginActivity.kt
package com.vald3nir.my_nanny.presentation.login

import android.os.Bundle
import android.view.inputmethod.EditorInfo
import androidx.lifecycle.Observer
import com.vald3nir.my_nanny.common.extensions.afterTextChanged
import com.vald3nir.my_nanny.common.core.BaseActivity
import com.vald3nir.my_nanny.databinding.ActivityLoginBinding
import org.koin.androidx.viewmodel.ext.android.viewModel

class LoginActivity : BaseActivity() {

    private val viewModel: LoginViewModel by viewModel()
    private lateinit var binding: ActivityLoginBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityLoginBinding.inflate(layoutInflater)
        setContentView(binding.root)
        viewModel.view = this
        setupObservers()
    }

    private fun setupObservers() {

        binding.apply {

            btnLogin.setOnClickListener { login() }
            btnRegister.setOnClickListener { register() }

            edtEmail.afterTextChanged {
                viewModel.loginDataChanged(
                    edtEmail.text.toString(),
                    edtPassword.text.toString()
                )
            }

            edtPassword.apply {
                afterTextChanged {
                    viewModel.loginDataChanged(
                        edtEmail.text.toString(),
                        edtPassword.text.toString()
                    )
                }

                setOnEditorActionListener { _, actionId, _ ->
                    when (actionId) {
                        EditorInfo.IME_ACTION_DONE ->
                            login()
                    }
                    false
                }
            }
        }

        viewModel.loginFormState.observe(this@LoginActivity, Observer {
            val loginState = it ?: return@Observer
            binding.btnLogin.isEnabled = loginState.isDataValid

            if (loginState.usernameError != null) {
                binding.edtEmail.error = getString(loginState.usernameError)
            }
            if (loginState.passwordError != null) {
                binding.edtPassword.error = getString(loginState.passwordError)
            }
        })
    }

    private fun ActivityLoginBinding.login() {
        viewModel.login(
            email = edtEmail.text.toString(),
            password = edtPassword.text.toString()
        )
    }

    private fun ActivityLoginBinding.register() {
        viewModel.register()
    }
}
